From d214f2a8625ca0c7a693cc46ac8136caec4f3de3 Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Wed, 09 Sep 2020 12:37:08 +0300
Subject: [PATCH] [SW-12754] gaudi: security: add n_if s_if protect

Add SOUTH_IF and NORTH_IF protection using
protection bits.

Change-Id: I198f850ec12f9b426dbce46373bb839d066131b6
Signed-off-by: Ran Shalit <rshalit@habana.ai>
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index d1dfed3..da64035 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -23,6 +23,8 @@
 #include <nic0_txe0_regs.h>
 #include <nic0_tmr_regs.h>
 #include <sif_rtr_ctrl_7_regs.h>
+#include <nif_rtr_0_regs.h>
+#include <nif_rtr_ctrl_0_regs.h>
 #include <psoc_global_conf_regs.h>
 
 #include <cpu_if_masks.h>
@@ -372,6 +374,142 @@
 	/* NIC2_HBM, NIC2_MME, NIC2_TPC */
 	base = mmNIC2_HBM_PLL_BASE & ~HL_CFG_BASE;
 	set_prot_bits_pages_range_all(base, 0x1000, 3);
+}
+
+static void preboot_secure_set_prot_bits_if_rtr(u32 base)
+{
+	u32 mask, prot_base = base;
+
+	/* offsets: 0 - 0x400 */
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 8);
+
+	/* offsets: 0x400 - 0x500 - allowed access */
+
+	/* offsets: 0x500 - 0x580 */
+	prot_base += SNGL_REG_PROT_SIZE * 10;
+	mask = ~(PROT_BIT(mmNIF_RTR_0_LBW_RANGE_PROT_MAX_AR_8) &
+		 PROT_BIT(mmNIF_RTR_0_LBW_RANGE_PROT_MAX_AR_9) &
+		 PROT_BIT(mmNIF_RTR_0_LBW_RANGE_PROT_MAX_AR_10) &
+		 PROT_BIT(mmNIF_RTR_0_LBW_RANGE_PROT_MAX_AR_11) &
+		 PROT_BIT(mmNIF_RTR_0_LBW_RANGE_PROT_MAX_AR_12) &
+		 PROT_BIT(mmNIF_RTR_0_LBW_RANGE_PROT_MAX_AR_13) &
+		 PROT_BIT(mmNIF_RTR_0_LBW_RANGE_PROT_MAX_AR_14) &
+		 PROT_BIT(mmNIF_RTR_0_LBW_RANGE_PROT_MAX_AR_15));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0x580 - 0xF80 */
+	prot_base += SNGL_REG_PROT_SIZE;
+	set_prot_bits_range_all(prot_base, PROT_SIZE_MAX - (prot_base - base));
+}
+
+static void preboot_secure_set_prot_bits_if_rtr_ctrl(u32 base)
+{
+	u32 mask, prot_base = base;
+
+	/* offsets: 0 - 0x580 */
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 11);
+
+	/* offsets: 0x580 - 0x600 */
+	prot_base += SNGL_REG_PROT_SIZE * 11;
+	mask = ~(PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AW_0) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AW_1) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AW_2) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AW_3) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AW_4) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AW_5) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AW_6));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0x600 - 0x680 allowed access */
+
+	/* offsets: 0x680 - 0x700 as following: */
+	prot_base += SNGL_REG_PROT_SIZE * 2;
+	mask = (PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AW_0) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AW_1) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AW_2) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AW_3) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AW_4) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AW_5) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AW_6));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0x700 - 0x800 not allowed access */
+	prot_base += SNGL_REG_PROT_SIZE;
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 2);
+
+	/* offsets: 0x800 - 0880 as following: */
+	prot_base += SNGL_REG_PROT_SIZE * 2;
+	mask =  ~(PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_0) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_1) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_2) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_3) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_4) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_5) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_6) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_7) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_8) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_9) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_10) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_11) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_12) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_13) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_14) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_LOW_AR_15) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_HIGH_AR_0) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_HIGH_AR_1) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_HIGH_AR_2) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_HIGH_AR_3) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_HIGH_AR_4) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_HIGH_AR_5) &
+		 PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_BASE_HIGH_AR_6));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0x880 - 0x900 allowed access */
+
+	/* offsets: 0x900 - 0x980 as following: */
+	prot_base += SNGL_REG_PROT_SIZE * 2;
+	mask = ~(PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_7) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_8) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_9) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_10) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_11) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_12) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_13) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_14) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_MASK_HIGH_AR_15));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0x980 - 0xA00 not allowed access */
+	prot_base += SNGL_REG_PROT_SIZE;
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE);
+
+	/* offsets: 0xA00 - 0xA80 as following: */
+	prot_base += SNGL_REG_PROT_SIZE;
+	mask = ~(PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_HIT_AW) &
+		PROT_BIT(mmNIF_RTR_CTRL_0_RANGE_SEC_HIT_AR));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0xA80 - 0xF80 */
+	prot_base += SNGL_REG_PROT_SIZE;
+	set_prot_bits_range_all(prot_base, PROT_SIZE_MAX - (prot_base - base));
+}
+
+#define IF_RTRS_DIST		0x10000
+#define IF_RTR_TO_RTR_CTRL_DIST	0x6000
+#define IF_RTRS_NUM		16
+
+static void preboot_secure_if(void)
+{
+	int i;
+	u32 prot_base, base = mmSIF_RTR_0_BASE & ~HL_CFG_BASE;
+
+	for (i = 0; i < IF_RTRS_NUM; i++) {
+		prot_base = base + i * IF_RTRS_DIST;
+		preboot_secure_set_prot_bits_if_rtr(prot_base);
+
+		prot_base += IF_RTR_TO_RTR_CTRL_DIST;
+		preboot_secure_set_prot_bits_if_rtr_ctrl(prot_base);
+	}
 }
 
 static void preboot_secure_psoc_global_conf(void)
@@ -734,6 +872,7 @@
 	preboot_secure_tpc();
 	preboot_secure_pll();
 	preboot_secure_nic();
+	preboot_secure_if();
 
 	/* TODO: implement the rest of security enablement */
 
